### PR TITLE
Added ValueKindDesc(kind) providing a string description for a value kind, for use in error messages and during debugging.

### PR DESCRIPTION
**Pull request by pk19604014**

### CODE REVIEW

File: explorer/interpreter/value.cpp [hunk 1]
```diff
@@ -524,6 +524,95 @@ void Value::Print(llvm::raw_ostream& out) const {
   }
 }
 
+auto ValueKindName(Value::Kind kind) -> std::string_view {
```
Comment (jonmeow) [inline explorer/interpreter/value.cpp:550]: If you want something like this, I'd suggest looking at what's being done on the toolchain side for enum stringification, or just printing the integer value. I don't think that we should maintain this kind of list.

https://github.com/carbon-language/carbon-lang/blob/trunk/toolchain/parser/parse_node_kind.h

**Review by jonmeow — CHANGES_REQUESTED** [review]

**Review by josh11b — APPROVED** [review]

File: explorer/interpreter/value.cpp (hunk 1 above)
Comment (pk19604014) [reply explorer/interpreter/value.cpp:550]: > If you want something like this, I'd suggest looking at what's being done on the toolchain side for enum stringification, or just printing the integer value. I don't think that we should maintain this kind of list.
> 
> https://github.com/carbon-language/carbon-lang/blob/trunk/toolchain/parser/parse_node_kind.h

Cool! Added value_kind.def and change enum definition and `ValueKindName` to use the new mechanism.

File: explorer/interpreter/value.cpp [hunk 2]
```diff
@@ -524,6 +525,22 @@ void Value::Print(llvm::raw_ostream& out) const {
   }
 }
 
+// Generates a 'human readable' description for an enum name.
+// For example, IntType -> 'int type'.
+static auto MakeKindName(const char* name) -> std::string {
+  return llvm::StringRef(
+             std::regex_replace(name, std::regex("(?!^)([A-Z])"), " $1"))
+      .lower();
```
Comment (jonmeow) [inline explorer/interpreter/value.cpp:563]: Can you elaborate why you aren't just returning the name directly? Doesn't this make it harder to figure out the enum value?

Comment (pk19604014) [reply explorer/interpreter/value.cpp:563]: > Can you elaborate why you aren't just returning the name directly? Doesn't this make it harder to figure out the enum value?

It does a little bit, but it also results in better looking error messages:
`Expected a type, but got parameterized entity name: Point` vs 
`Expected a type, but got ParameterizedEntityName: Point`

And also matches the behavior of `{Expression/Pattern/Statement/Declaration}KindName()`

Comment (jonmeow) [reply explorer/interpreter/value.cpp:563]: I'll look for someone else to comment -- for me, I'd rather see that it's an enum value and have something I can search on, rather than something I can't.

File: explorer/interpreter/value_kind.def [hunk 3]
```diff
@@ -0,0 +1,59 @@
+// Part of the Carbon Language project) under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// Note that this is an X-macro header.
+//
+// It does not use `#include` guards) and instead is designed to be `#include`ed
```
Comment (josh11b) [inline explorer/interpreter/value_kind.def]: Looks like some search replace `,` -> `)` went too far

Comment (pk19604014) [reply explorer/interpreter/value_kind.def]: > Looks like some search replace `,` -> `)` went too far

Whoops, fixed.

Btw, is there a way to make git[hub] indicate that this file is a copy of value.h to better preserve history?

File: explorer/interpreter/value.cpp (hunk 2 above)
Comment (pk19604014) [reply explorer/interpreter/value.cpp:563]: > I'll look for someone else to comment -- for me, I'd rather see that it's an enum value and have something I can search on, rather than something I can't.

Can someone look into this?

Comment (josh11b) [reply explorer/interpreter/value.cpp:563]: @geoffromer Could you look at this when you get back tomorrow?

Comment (jsiek) [reply explorer/interpreter/value.cpp:563]: In general, I'm in favor of programmer-facing errors being written in plain English. (On the other hand, internal errors can and should explicitly mention internal components of the explorer.)

Comment (geoffromer) [reply explorer/interpreter/value.cpp:563]: (Sorry, was out of the office for longer than expected)

As @jsiek points out, there are use cases for both getting the name of a given enumerator, and getting a prose description of the kind of value it represents, so we should probably leave room for both in the API. In particular, it seems to me that the _name_ of `IntVaue` is "IntValue", not "int value", so a function called `ValueKindName` that returns "int value" seems liable to cause confusion. I would recommend renaming this function to something like `ValueKindDescription`, or changing it to return the actual enumerator names (or both).

Do I understand correctly that the current approach will produce error messages like these?

```
Expected a type, but got int value: ...
Expected a type, but got symbolic witness: ...
Expected a type, but got l value: ...
Expected a type, but got addr value: ...
```

These seem like the worst of both worlds: they don't give an enumerator name I can search for in the explorer code, but they also aren't very clear (or even grammatical) as prose. If we want to strive for clear prose error messages, I'd want to see something more like:

```
Expected a type, but got an integer value: ...
Expected a type, but got a symbolic witness: ...
Expected a type, but got an l-value: ...
Expected a type, but got an `addr` pattern value: ...
```

But we can't get there by applying a regex to the enumerator names. Conversely, the regex approach will constrain how we name these types, because the best type name in C++ code might not rewrite to the best description to use in an error message. To do this properly, we'd need something like a data file that maps type names to their prose descriptions. I'm not sure even that would be sufficient, because in other contexts we might not want the "a"/"an" prefix. And enumerators might not be the right lookup keys, because we might use multiple C++ types to represent the same user-facing concept, or vice-versa.

My recommendation would be to postpone diving down that rathole, and stick with enumerator names for now.

> And also matches the behavior of `{Expression/Pattern/Statement/Declaration}KindName()`

I agree they should be consistent, but I think the points above about `ValueKindName` also apply to these functions, so I'd prefer to have those functions also return the enumerator name unchanged.

Comment (pk19604014) [reply explorer/interpreter/value.cpp:563]: I really don't like raw enumerator names in some contexts like the error messages, so maybe the solution is to have both?

For this PR, renamed XxKindName() -> XxKindDesc() both for the new and existing functions, to keep consistency. 
PTAL ;)

File: explorer/interpreter/expression_checker.cpp [hunk 4]
```diff
@@ -0,0 +1,1175 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
```
Comment (josh11b) [inline explorer/interpreter/expression_checker.cpp]: Is this supposed to be added in this PR?

Comment (pk19604014) [reply explorer/interpreter/expression_checker.cpp]: > Is this supposed to be added in this PR?

Whoops, fixed

File: explorer/interpreter/value.cpp (hunk 2 above)
Comment (jonmeow) [reply explorer/interpreter/value.cpp:563]: I'd rather just have the enum name rather than trying to ensure that a regex replacement of the enum name yields a reasonable string. I think the enum name should remain human readable.

Comment (pk19604014) [reply explorer/interpreter/value.cpp:563]: > I'd rather just have the enum name rather than trying to ensure that a regex replacement of the enum name yields a reasonable string. I think the enum name should remain human readable.

The string-splitting does seem to work well enough though, and I like it more. I can add XxKindName() for all impacted enums in a follow-up PR.

Comment (jonmeow) [reply explorer/interpreter/value.cpp:563]: I think if you want to go back to the initial approach, without the `.def` because it's then just adding unnecessary complexity, and provide the more readable text that @geoffromer suggested, I'd be fine with that (since I think you and he will have agreed on it).

If you wanted to stick with the enum, I'd also be fine with that -- it's the approach I personally favor.

However, we should work towards an approach that at least some majority agrees with.

Comment (pk19604014) [reply explorer/interpreter/value.cpp:563]: > I think if you want to go back to the initial approach, without the `.def` because it's then just adding unnecessary complexity, and provide the more readable text that @geoffromer suggested, I'd be fine with that (since I think you and he will have agreed on it).
> 
> If you wanted to stick with the enum, I'd also be fine with that -- it's the approach I personally favor.
> 
> However, we should work towards an approach that at least some majority agrees with.

Sure I maybe could do one of these, but given the PR is merely adding a simple debug output aid and not like trying to solve some complicated language design issue, is it even worth spending so much time (18 days so far for this tiny PR) on it debating the space character? Is it ok to have a non 100% ideal change (from someone's point of view anyways) and then evaluate and fix if it's really not working out, in the launch-and-iterate spirit?

I actually am now quite reluctant to start doing other/bigger PRs because I feel they might take forever because of some similar minor/arbitrary thing. @chandlerc for thoughts.

Comment (chandlerc) [reply explorer/interpreter/value.cpp:563]: FWIW, I agree with the direction suggested by @geoffromer (and originally by @jonmeow):

> My recommendation would be to postpone diving down that rathole, and stick with enumerator names for now.

I really do understand that you don't like the resulting error messages. I agree they could be better for users. But I think that is better addressed with a holistic design, and a design that is discussed with folks working on the explorer ahead of time. As several folks have pointed out, using a regular expression to parse the enum name is going to quickly hit issues.

To get good user-facing error strings, we're going to need some form of proper diagnostics engine that formats errors in a human readable way. Clang and even bits of the toolchain can serve as ideas for these designs, but I think they will look substantially different and be a substantially larger change.

Geoff's point is to not try to solve that problem today (in this PR), and use the much simpler form of just printing the actual enum as it is spelled in the explorer codebase. That makes sense to me given the current state of things.

However, you also raised an important point:

> And also matches the behavior of {Expression/Pattern/Statement/Declaration}KindName()

But @geoffromer directly suggested those should change to also directly print the enum. IMO, that also makes sense.

So I think there are two clear paths forward here that maybe haven't been stated as explicitly as they might:

- a) Land this PR with the exact spelling of the enums. Send a follow-up PR to change the other `*KindName()` functions to be consistent.
- b) Land this PR with the spelling consistent with existing `*KindName()` functions. Send a follow-up PR to change all the `*KindName()` functions to directly return the enum names.

It seems like (b) is strictly more work than (a) so I would probably suggest (a). I think that was the implicit suggestion from both Jon and Geoff.

---

> Sure I maybe could do one of these, but given the PR is merely adding a simple debug output aid and not like trying to solve some complicated language design issue, is it even worth spending so much time (18 days so far for this tiny PR) on it debating the space character? Is it ok to have a non 100% ideal change (from someone's point of view anyways) and then evaluate and fix if it's really not working out, in the launch-and-iterate spirit?

No one is suggesting to have a 100% ideal change, they're just suggesting to avoid adding another instance of complexity that doesn't make sense to them as a code reviewer. I think there is a pretty clear path forward here.

> I actually am now quite reluctant to start doing other/bigger PRs because I feel they might take forever because of some similar minor/arbitrary thing. @chandlerc for thoughts.

I'm sorry you're reluctant, but I do think there was already a pretty clear path forward here.

I think I've mentioned this before, but I feel like discussing planned changes ahead of time, or discussing disagreements in real time in Discord or one of the meetings might help speed things up. Moreover, I think if you want to make more incremental / fast progress, it may help to move longer discussions around changing the direction suggested in code review to not be in the critical path of the code review and instead be separate discussions.

Anyways, my two cents here.

File: explorer/interpreter/value_kind.def [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+// It does not use `#include` guards, and instead is designed to be `#include`ed
+// after the x-macro is defined in order for its inclusion to expand to the
+// desired output. The x-macro for this header is `CARBON_VALUE_KIND`. The
+// definition provided will be removed at the end of this file to clean up.
+
+#ifndef CARBON_VALUE_KIND
+#error "Must define the x-macro to use this file."
+#endif
+
+CARBON_VALUE_KIND(IntValue)
```
Comment (josh11b) [inline explorer/interpreter/value_kind.def:16]: And idea would be to change this to `CARBON_VALUE_KIND(IntValue, "an integer value")`, so the string is customizable and include the article. Would that address the concerns here?

Comment (chandlerc) [reply explorer/interpreter/value_kind.def:16]: FWIW, my experience with Clang makes me skeptical about how well this will work long term. I think we'll need something fairly similar to Clang's formatted rendering if messages at least. You can see a fairly simple model for this in the Carbon toolchain/diagnostics library.

But while the primary users of the Carbon explorer are also very likely *developing* Carbon (in one form or another), it seems relatively reasonable to use something search-able in the codebase like the enum itself, and then work on building a more user-facing diagnostics system when we're expecting to have a more broad userbase. Right now, it feels a bit like we're investing time trying to build "good" user facing messages when that's not the most relevant thing.

But on the flip side, getting this exactly "right" also doesn't seem critical. I was mostly chiming in to try to help get *some* path forward here...

File: explorer/interpreter/value.cpp (hunk 2 above)
Comment (pk19604014) [reply explorer/interpreter/value.cpp:563]: > FWIW, I agree with the direction suggested by @geoffromer (and originally by @jonmeow):
> 
> > My recommendation would be to postpone diving down that rathole, and stick with enumerator names for now.
> 
> I really do understand that you don't like the resulting error messages. I agree they could be better for users. But I think that is better addressed with a holistic design, and a design that is discussed with folks working on the explorer ahead of time. As several folks have pointed out, using a regular expression to parse the enum name is going to quickly hit issues.
> 
> To get good user-facing error strings, we're going to need some form of proper diagnostics engine that formats errors in a human readable way. Clang and even bits of the toolchain can serve as ideas for these designs, but I think they will look substantially different and be a substantially larger change.
> 
> Geoff's point is to not try to solve that problem today (in this PR), and use the much simpler form of just printing the actual enum as it is spelled in the explorer codebase. That makes sense to me given the current state of things.
> 
> However, you also raised an important point:
> 
> > And also matches the behavior of {Expression/Pattern/Statement/Declaration}KindName()
> 
> But @geoffromer directly suggested those should change to also directly print the enum. IMO, that also makes sense.
> 
> So I think there are two clear paths forward here that maybe haven't been stated as explicitly as they might:
> 
> * a) Land this PR with the exact spelling of the enums. Send a follow-up PR to change the other `*KindName()` functions to be consistent.
> * b) Land this PR with the spelling consistent with existing `*KindName()` functions. Send a follow-up PR to change all the `*KindName()` functions to directly return the enum names.
> 
> It seems like (b) is strictly more work than (a) so I would probably suggest (a). I think that was the implicit suggestion from both Jon and Geoff.
> 
> > Sure I maybe could do one of these, but given the PR is merely adding a simple debug output aid and not like trying to solve some complicated language design issue, is it even worth spending so much time (18 days so far for this tiny PR) on it debating the space character? Is it ok to have a non 100% ideal change (from someone's point of view anyways) and then evaluate and fix if it's really not working out, in the launch-and-iterate spirit?
> 
> No one is suggesting to have a 100% ideal change, they're just suggesting to avoid adding another instance of complexity that doesn't make sense to them as a code reviewer. I think there is a pretty clear path forward here.
> 
> > I actually am now quite reluctant to start doing other/bigger PRs because I feel they might take forever because of some similar minor/arbitrary thing. @chandlerc for thoughts.
> 
> I'm sorry you're reluctant, but I do think there was already a pretty clear path forward here.
> 
> I think I've mentioned this before, but I feel like discussing planned changes ahead of time, or discussing disagreements in real time in Discord or one of the meetings might help speed things up. Moreover, I think if you want to make more incremental / fast progress, it may help to move longer discussions around changing the direction suggested in code review to not be in the critical path of the code review and instead be separate discussions.
> 
> Anyways, my two cents here.

Thanks Chandler. I guess my comment was more high level -- every single case usually has a 'way forward', but this is not the first time where a seemingly trivial PR I write results in a prolonged discussion which impacts velocity quite a bit, so I find it hard to justify based on an effort-vs-impact (and risk) basis. But maybe the approach here is different than what I am used to, or I'm misclassifying things as trivial whereas they are not? Not sure ;)

I went ahead and added a MakeKindName() method using the same .def machinery.
